File: cmd/triage.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os/exec"
	"strings"
	"time"

	"github.com/mturley/agent-handler/config"
	"github.com/mturley/agent-handler/discover"
	"github.com/spf13/cobra"
)

var triageCmd = &cobra.Command{
	Use:   "triage",
	Short: "Aggregate what needs attention across all sessions",
	RunE:  runTriage,
}

func init() {
	triageCmd.GroupID = "human"
	rootCmd.AddCommand(triageCmd)
}

type triageOutput struct {
	SessionsActive        int                   `json:"sessions_active"`
	SessionsBlocked       int                   `json:"sessions_blocked"`
	SessionsDead          int                   `json:"sessions_dead"`
	BlockedSessions       []blockedSession      `json:"blocked_sessions"`
	SessionsWithUnread    []sessionUnread       `json:"sessions_with_unread"`
	WatcherErrors         []watcherError        `json:"watcher_errors"`
	EventsSinceLastCheck  int                   `json:"events_since_last_check"`
	DeadSessions          []deadSession         `json:"dead_sessions"`
	SessionResources      []sessionResource     `json:"session_resources"`
	StaleResources        []staleResource       `json:"stale_resources"`
}

type blockedSession struct {
	SessionID    string `json:"session_id"`
	SessionName  string `json:"session_name"`
	Branch       string `json:"branch"`
	BlockedSince string `json:"blocked_since"`
}

type sessionUnread struct {
	SessionID    string         `json:"session_id"`
	SessionName  string         `json:"session_name"`
	UnreadCount  int            `json:"unread_count"`
	UnreadTypes  map[string]int `json:"unread_types"`
}

type watcherError struct {
	Name             string `json:"name"`
	LastErrorMessage string `json:"last_error_message"`
}

type deadSession struct {
	SessionID  string `json:"session_id"`
	SessionName string `json:"session_name"`
	LastActive string `json:"last_active"`
}

type sessionResource struct {
	SessionID    string            `json:"session_id"`
	SessionName  string            `json:"session_name"`
	Resources    []resourceDetail  `json:"resources"`
}

type resourceDetail struct {
	ResourceType     string          `json:"resource_type"`
	ResourceID       string          `json:"resource_id"`
	ResourceURL      *string         `json:"resource_url,omitempty"`
	State            json.RawMessage `json:"state"`
	WatcherUpdatedAt string          `json:"watcher_updated_at,omitempty"`
}

type staleResource struct {
	ResourceType     string `json:"resource_type"`
	ResourceID       string `json:"resource_id"`
	WatcherUpdatedAt string `json:"watcher_updated_at"`
	StaleMinutes     int    `json:"stale_minutes"`
}

func runTriage(cmd *cobra.Command, args []string) error {
	d, err := openReadOnlyDB()
	if err != nil {
		return fmt.Errorf("failed to open database: %w", err)
	}
	defer d.Close()

	output := triageOutput{
		BlockedSessions:    []blockedSession{},
		SessionsWithUnread: []sessionUnread{},
		WatcherErrors:      []watcherError{},
		DeadSessions:       []deadSession{},
		SessionResources:   []sessionResource{},
		StaleResources:     []staleResource{},
	}

	// Get all active sessions
	sessions, err := d.ListSessions(false, 1000, 0)
	if err != nil {
		return fmt.Errorf("failed to list sessions: %w", err)
	}

	// Count active sessions
	for _, s := range sessions {
		if s.Status == "active" {
			output.SessionsActive++
		}
	}

	// Find blocked sessions
	rows, err := d.Query(`
		SELECT s.session_id, COALESCE(s.session_name,''), s.branch, e.ts as blocked_since
		FROM sessions s
		JOIN events e ON e.session_id = s.session_id AND e.type = 'blocked'
		WHERE s.status = 'active'
		  AND NOT EXISTS (
		    SELECT 1 FROM events e2
		    WHERE e2.session_id = s.session_id AND e2.type = 'unblocked' AND e2.ts > e.ts
		  )
	`)
	if err != nil {
		return fmt.Errorf("failed to query blocked sessions: %w", err)
	}
	defer rows.Close()

	for rows.Next() {
		var bs blockedSession
		if err := rows.Scan(&bs.SessionID, &bs.SessionName, &bs.Branch, &bs.BlockedSince); err != nil {
			return fmt.Errorf("failed to scan blocked session: %w", err)
		}
		output.BlockedSessions = append(output.BlockedSessions, bs)
		output.SessionsBlocked++
	}

	if err := rows.Err(); err != nil {
		return fmt.Errorf("error iterating blocked sessions: %w", err)
	}

	// Find sessions with unread events
	for _, s := range sessions {
		if s.Status != "active" {
			continue
		}
		count, breakdown, err := d.UnreadCountForSession(s.SessionID)
		if err != nil {
			return fmt.Errorf("failed to get unread count for session %s: %w", s.SessionID, err)
		}
		if count > 0 {
			output.SessionsWithUnread = append(output.SessionsWithUnread, sessionUnread{
				SessionID:   s.SessionID,
				SessionName: s.SessionName,
				UnreadCount: count,
				UnreadTypes: breakdown,
			})
		}
	}

	// Check watcher errors
	for _, svc := range []string{"github", "jira"} {
		if d.HasWatcherError(svc) {
			ws, err := d.GetWatcherStatus(svc)
			if err != nil {
				return fmt.Errorf("failed to get watcher status for %s: %w", svc, err)
			}
			if ws != nil {
				output.WatcherErrors = append(output.WatcherErrors, watcherError{
					Name:             svc,
					LastErrorMessage: ws.LastErrorMessage,
				})
			}
		}
	}

	// Count events since last check (for the current session)
	currentSessionID, err := resolveSessionID(cmd)
	if err == nil && currentSessionID != "" {
		cursor, err := d.GetCursor(currentSessionID)
		if err == nil {
			if cursor == "" {
				cursor = "1970-01-01T00:00:00Z"
			}
			var count int
			err = d.QueryRow(`SELECT COUNT(*) FROM events WHERE ts > ?`, cursor).Scan(&count)
			if err == nil {
				output.EventsSinceLastCheck = count
			}
		}
	}

	// Find dead sessions and adjust active count
	for _, s := range sessions {
		if s.Status != "active" {
			continue
		}
		if s.PID > 0 && !discover.IsSessionProcess(s.PID, s.SessionID) {
			output.DeadSessions = append(output.DeadSessions, deadSession{
				SessionID:  s.SessionID,
				SessionName: s.SessionName,
				LastActive: s.LastActive,
			})
			output.SessionsDead++
			output.SessionsActive--
		}
	}

	// Gather resource state per session
	staleThreshold := 5 * time.Minute
	seenResources := make(map[string]bool) // dedup stale resources across sessions

	for _, s := range sessions {
		if s.Status != "active" {
			continue
		}
		// Skip dead sessions
		isDead := false
		for _, ds := range output.DeadSessions {
			if ds.SessionID == s.SessionID {
				isDead = true
				break
			}
		}
		if isDead {
			continue
		}

		states, err := d.ListResourceStatesForSession(s.SessionID)
		if err != nil || len(states) == 0 {
			continue
		}

		sr := sessionResource{
			SessionID:   s.SessionID,
			SessionName: s.SessionName,
			Resources:   []resourceDetail{},
		}

		for _, rs := range states {
			rd := resourceDetail{
				ResourceType:     rs.ResourceType,
				ResourceID:       rs.ResourceID,
				ResourceURL:      rs.ResourceURL,
				State:            json.RawMessage(rs.StateJSON),
				WatcherUpdatedAt: rs.WatcherUpdatedAt,
			}
			sr.Resources = append(sr.Resources, rd)

			// Check staleness
			key := rs.ResourceType + ":" + rs.ResourceID
			if !seenResources[key] && rs.WatcherUpdatedAt != "" {
				seenResources[key] = true
				wut, err := time.Parse(time.RFC3339, rs.WatcherUpdatedAt)
				if err == nil && time.Since(wut) > staleThreshold {
					output.StaleResources = append(output.StaleResources, staleResource{
						ResourceType:     rs.ResourceType,
						ResourceID:       rs.ResourceID,
						WatcherUpdatedAt: rs.WatcherUpdatedAt,
						StaleMinutes:     int(time.Since(wut).Minutes()),
					})
				}
			}
		}

		output.SessionResources = append(output.SessionResources, sr)
	}

	// Trigger catch-up for stale resources (best-effort, non-blocking)
	if len(output.StaleResources) > 0 {
		cfg, _ := config.Read(config.DefaultPath())
		if cfg != nil {
			staleByService := make(map[string][]string)
			for _, sr := range output.StaleResources {
				svc := config.ResourceTypeToService(sr.ResourceType)
				if svc != "" && cfg.IsServiceConfigured(svc) {
					staleByService[svc] = append(staleByService[svc], sr.ResourceID)
				}
			}
			for svc, resources := range staleByService {
				resourceList := strings.Join(resources, ",")
				go func(s, r string) {
					exec.Command("handler", "watcher", "run", s, "--resources", r).Run()
				}(svc, resourceList)
			}
		}
	}

	// Output
	if jsonOutput {
		enc := json.NewEncoder(cmd.OutOrStdout())
		enc.SetIndent("", "  ")
		return enc.Encode(output)
	}

	// Text output
	fmt.Printf("Handler Triage\n\n")

	fmt.Printf("Sessions: %d active", output.SessionsActive)
	if output.SessionsBlocked > 0 {
		fmt.Printf(", %d blocked", output.SessionsBlocked)
	}
	if output.SessionsDead > 0 {
		fmt.Printf(", %d dead", output.SessionsDead)
	}
	fmt.Println()

	if len(output.BlockedSessions) > 0 {
		fmt.Println("\nBlocked Sessions:")
		for _, bs := range output.BlockedSessions {
			name := bs.SessionName
			if name == "" {
				name = bs.SessionID[:8]
			}
			fmt.Printf("  %s (%s) - blocked since %s\n", name, bs.Branch, bs.BlockedSince)
		}
	}

	if len(output.SessionsWithUnread) > 0 {
		fmt.Println("\nSessions with Unread Events:")
		for _, su := range output.SessionsWithUnread {
			name := su.SessionName
			if name == "" {
				name = su.SessionID[:8]
			}
			typeBreakdown := ""
			if len(su.UnreadTypes) > 0 {
				var parts []string
				for t, c := range su.UnreadTypes {
					parts = append(parts, fmt.Sprintf("%s: %d", t, c))
				}
				typeBreakdown = fmt.Sprintf(" (%s)", strings.Join(parts, ", "))
			}
			fmt.Printf("  %s - %d unread%s\n", name, su.UnreadCount, typeBreakdown)
		}
	}

	if len(output.WatcherErrors) > 0 {
		fmt.Println("\nWatcher Errors:")
		for _, we := range output.WatcherErrors {
			fmt.Printf("  %s: %s\n", we.Name, we.LastErrorMessage)
		}
	}

	if output.EventsSinceLastCheck > 0 {
		fmt.Printf("\nNew Events: %d since last check\n", output.EventsSinceLastCheck)
	}

	if len(output.DeadSessions) > 0 {
		fmt.Println("\nDead Sessions:")
		for _, ds := range output.DeadSessions {
			name := ds.SessionName
			if name == "" {
				name = ds.SessionID[:8]
			}
			fmt.Printf("  %s - last active %s\n", name, ds.LastActive)
		}
	}

	if len(output.SessionResources) > 0 {
		fmt.Println("\nSession Resources:")
		for _, sr := range output.SessionResources {
			name := sr.SessionName
			if name == "" {
				name = sr.SessionID[:8]
			}
			for _, r := range sr.Resources {
				fmt.Printf("  %s → %s:%s", name, r.ResourceType, r.ResourceID)
				if r.WatcherUpdatedAt != "" {
					wut, err := time.Parse(time.RFC3339, r.WatcherUpdatedAt)
					if err == nil {
						fmt.Printf(" (updated %s ago)", formatDuration(time.Since(wut)))
					}
				}
				fmt.Println()
			}
		}
	}

	if len(output.StaleResources) > 0 {
		fmt.Println("\nStale Resources (catch-up triggered):")
		for _, sr := range output.StaleResources {
			fmt.Printf("  %s:%s — last updated %dm ago\n", sr.ResourceType, sr.ResourceID, sr.StaleMinutes)
		}
	}

	if len(output.BlockedSessions) == 0 && len(output.SessionsWithUnread) == 0 &&
		len(output.WatcherErrors) == 0 && len(output.DeadSessions) == 0 &&
		output.EventsSinceLastCheck == 0 {
		fmt.Println("\nAll clear - nothing needs attention.")
	}

	return nil
}

File: src/Handler/Cmd/Triage.hs
-- | Port of cmd/triage.go: aggregate what needs attention across sessions.
module Handler.Cmd.Triage (triageCommand) where

import Control.Monad (filterM, forM, forM_, unless, void, when)
import Data.Aeson (Value(..), decodeStrict, object, toJSON, (.=))
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Maybe (catMaybes, fromMaybe)
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE
import qualified Database.SQLite.Simple as SQL
import Options.Applicative
import System.Process (spawnProcess)

import Handler.Cli.Common
import Handler.Cmd.Status (formatDuration, parseIso, sinceSeconds)
import Handler.Config (configDefaultPath, isServiceConfigured, readConfig, resourceTypeToService)
import Handler.Db (Db(..), close, handlerHome)
import Handler.Db.Cursors (getCursor)
import Handler.Db.Events (unreadCountForSession)
import Handler.Db.ResourceState (ResourceStateWithSubscription(..), listResourceStatesForSession)
import Handler.Db.Sessions (Session(..), listSessions)
import Handler.Db.WatcherStatus (WatcherStatus(..), getWatcherStatus, hasWatcherError)
import Handler.Discover (isSessionProcess, resolveSessionId)
import Handler.Util (epochIso)

triageCommand :: Mod CommandFields NamedCommand
triageCommand = mkCommand "triage" "Aggregate what needs attention across all sessions" (pure runTriage)

data StaleResource = StaleResource
  { srType      :: Text
  , srId        :: Text
  , srUpdatedAt :: Text
  , srMinutes   :: Int
  }

runTriage :: Ctx -> IO ()
runTriage ctx = do
  db <- openReadOnlyDb
  sessions <- listSessions db False 1000 0

  -- Blocked sessions: a 'blocked' event with no later 'unblocked'
  blockedRows <- SQL.query_ db.conn
    "SELECT s.session_id, COALESCE(s.session_name,''), s.branch, e.ts as blocked_since\
    \ FROM sessions s\
    \ JOIN events e ON e.session_id = s.session_id AND e.type = 'blocked'\
    \ WHERE s.status = 'active'\
    \   AND NOT EXISTS (\
    \     SELECT 1 FROM events e2\
    \     WHERE e2.session_id = s.session_id AND e2.type = 'unblocked' AND e2.ts > e.ts\
    \   )"
    :: IO [(Text, Text, Text, Text)]

  -- Sessions with unread events
  sessionsWithUnread <- fmap catMaybes $ forM sessions $ \s ->
    if s.status /= "active"
      then pure Nothing
      else do
        (count, breakdown) <- unreadCountForSession db s.sessionId
        pure $ if count > 0 then Just (s, count, breakdown) else Nothing

  -- Watcher errors
  watcherErrors <- fmap catMaybes $ forM (["github", "jira"] :: [Text]) $ \svc -> do
    hasErr <- hasWatcherError db svc
    if not hasErr
      then pure Nothing
      else do
        mws <- getWatcherStatus db svc
        pure $ (\ws -> (svc, ws.lastErrorMessage)) <$> mws

  -- Events since last check (for the current session; best-effort)
  eventsSinceLastCheck :: Int <- do
    msid <- bestEffortSessionId
    case msid of
      Nothing -> pure 0
      Just sid -> do
        cursor0 <- getCursor db sid
        let cursor = if cursor0 == "" then epochIso else cursor0
        counts <- SQL.query db.conn "SELECT COUNT(*) FROM events WHERE ts > ?" (SQL.Only cursor)
        pure $ case counts of { (SQL.Only n : _) -> n; _ -> 0 }

  -- Dead sessions
  deadSessions <- flip filterM sessions $ \s ->
    if s.status /= "active"
      then pure False
      else if s.pid > 0
        then not <$> isSessionProcess s.pid s.sessionId
        else pure False
  let deadIds = [s.sessionId | s <- deadSessions]
      sessionsActive = length [s | s <- sessions, s.status == "active"] - length deadSessions
      sessionsBlocked = length blockedRows
      sessionsDead = length deadSessions

  -- Resource state per session + staleness (5 min threshold, deduped)
  let staleThreshold = 5 * 60 :: Double
  (sessionResources, staleResources) <- do
    resultsAndStales <- forM [s | s <- sessions, s.status == "active", s.sessionId `notElem` deadIds] $ \s -> do
      states <- listResourceStatesForSession db s.sessionId
      if null states
        then pure (Nothing, [])
        else do
          stales <- fmap catMaybes $ forM states $ \rs ->
            if rs.watcherUpdatedAt == ""
              then pure Nothing
              else case parseIso rs.watcherUpdatedAt of
                Nothing -> pure Nothing
                Just wut -> do
                  since <- sinceSeconds wut
                  pure $ if since > staleThreshold
                    then Just (StaleResource rs.resourceType rs.resourceId rs.watcherUpdatedAt (truncate (since / 60)))
                    else Nothing
          pure (Just (s, states), stales)
    let sessionResources = catMaybes (map fst resultsAndStales)
        -- dedup stale resources across sessions, preserving order
        allStales = concatMap snd resultsAndStales
        dedupStales seen = \case
          [] -> []
          (sr : rest)
            | (sr.srType <> ":" <> sr.srId) `elem` seen -> dedupStales seen rest
            | otherwise -> sr : dedupStales ((sr.srType <> ":" <> sr.srId) : seen) rest
    pure (sessionResources, dedupStales [] allStales)

  -- Trigger catch-up for stale resources (best-effort, non-blocking)
  unless (null staleResources) $ do
    cfg <- configDefaultPath >>= readConfig
    let byService = Map.fromListWith (flip (++))
          [ (svc, [sr.srId])
          | sr <- staleResources
          , let svc = resourceTypeToService sr.srType
          , svc /= ""
          , isServiceConfigured cfg svc
          ]
    forM_ (Map.toList byService) $ \(svc, resources) ->
      void $ spawnProcess "handler"
        ["watcher", "run", T.unpack svc, "--resources", T.unpack (T.intercalate "," resources)]

  if ctx.jsonOutput
    then printJson $ object
      [ "sessions_active" .= sessionsActive
      , "sessions_blocked" .= sessionsBlocked
      , "sessions_dead" .= sessionsDead
      , "blocked_sessions" .= toJSON
          [ object [ "session_id" .= sid, "session_name" .= name
                   , "branch" .= branch, "blocked_since" .= since ]
          | (sid, name, branch, since) <- blockedRows ]
      , "sessions_with_unread" .= toJSON
          [ object [ "session_id" .= s.sessionId, "session_name" .= s.sessionName
                   , "unread_count" .= count, "unread_types" .= breakdown ]
          | (s, count, breakdown) <- sessionsWithUnread ]
      , "watcher_errors" .= toJSON
          [ object [ "name" .= n, "last_error_message" .= m ] | (n, m) <- watcherErrors ]
      , "events_since_last_check" .= eventsSinceLastCheck
      , "dead_sessions" .= toJSON
          [ object [ "session_id" .= s.sessionId, "session_name" .= s.sessionName
                   , "last_active" .= s.lastActive ]
          | s <- deadSessions ]
      , "session_resources" .= toJSON
          [ object [ "session_id" .= s.sessionId, "session_name" .= s.sessionName
                   , "resources" .= toJSON (map resourceDetailJson states) ]
          | (s, states) <- sessionResources ]
      , "stale_resources" .= toJSON
          [ object [ "resource_type" .= sr.srType, "resource_id" .= sr.srId
                   , "watcher_updated_at" .= sr.srUpdatedAt, "stale_minutes" .= sr.srMinutes ]
          | sr <- staleResources ]
      ]
    else do
      putTextLn "Handler Triage\n"

      putTextLn $ "Sessions: " <> T.pack (show sessionsActive) <> " active"
        <> (if sessionsBlocked > 0 then ", " <> T.pack (show sessionsBlocked) <> " blocked" else "")
        <> (if sessionsDead > 0 then ", " <> T.pack (show sessionsDead) <> " dead" else "")

      unless (null blockedRows) $ do
        putTextLn "\nBlocked Sessions:"
        forM_ blockedRows $ \(sid, name0, branch, since) -> do
          let name = if name0 == "" then T.take 8 sid else name0
          printfT ["  ", name, " (", branch, ") - blocked since ", since]

      unless (null sessionsWithUnread) $ do
        putTextLn "\nSessions with Unread Events:"
        forM_ sessionsWithUnread $ \(s, count, breakdown) -> do
          let name = if s.sessionName == "" then T.take 8 s.sessionId else s.sessionName
              typeBreakdown = if Map.null breakdown then ""
                else " (" <> T.intercalate ", "
                       [t <> ": " <> T.pack (show c) | (t, c) <- Map.toList breakdown] <> ")"
          printfT ["  ", name, " - ", T.pack (show count), " unread", typeBreakdown]

      unless (null watcherErrors) $ do
        putTextLn "\nWatcher Errors:"
        forM_ watcherErrors $ \(n, m) -> printfT ["  ", n, ": ", m]

      when (eventsSinceLastCheck > 0) $
        printfT ["\nNew Events: ", T.pack (show eventsSinceLastCheck), " since last check"]

      unless (null deadSessions) $ do
        putTextLn "\nDead Sessions:"
        forM_ deadSessions $ \s -> do
          let name = if s.sessionName == "" then T.take 8 s.sessionId else s.sessionName
          printfT ["  ", name, " - last active ", s.lastActive]

      unless (null sessionResources) $ do
        putTextLn "\nSession Resources:"
        forM_ sessionResources $ \(s, states) -> do
          let name = if s.sessionName == "" then T.take 8 s.sessionId else s.sessionName
          forM_ states $ \rs -> do
            suffix <- case parseIso rs.watcherUpdatedAt of
              Just wut | rs.watcherUpdatedAt /= "" -> do
                since <- sinceSeconds wut
                pure (" (updated " <> formatDuration since <> " ago)")
              _ -> pure ""
            printfT ["  ", name, " → ", rs.resourceType, ":", rs.resourceId, suffix]

      unless (null staleResources) $ do
        putTextLn "\nStale Resources (catch-up triggered):"
        forM_ staleResources $ \sr ->
          printfT ["  ", sr.srType, ":", sr.srId, " — last updated ", T.pack (show sr.srMinutes), "m ago"]

      when (null blockedRows && null sessionsWithUnread && null watcherErrors
            && null deadSessions && eventsSinceLastCheck == 0) $
        putTextLn "\nAll clear - nothing needs attention."
  close db
  where
    resourceDetailJson :: ResourceStateWithSubscription -> Value
    resourceDetailJson rs = object $
      [ "resource_type" .= rs.resourceType
      , "resource_id" .= rs.resourceId
      , "state" .= stateValue
      ]
      ++ [ "resource_url" .= u | Just u <- [rs.resourceUrl] ]
      ++ [ "watcher_updated_at" .= rs.watcherUpdatedAt | rs.watcherUpdatedAt /= "" ]
      where
        stateValue = fromMaybe (String rs.stateJson) (decodeStrict (TE.encodeUtf8 rs.stateJson))

    -- resolveSessionIdOpt dies on failure; triage treats failure as "no session"
    bestEffortSessionId :: IO (Maybe Text)
    bestEffortSessionId = do
      home <- handlerHome
      either (const Nothing) Just <$> resolveSessionId home
